vocode/streaming/transcriber/deepgram_transcriber.py

```diff
@@ -17,55 +18,122 @@
     DeepgramTranscriberConfig,
     PunctuationEndpointingConfig,
     TimeEndpointingConfig,
+    TranscriberConfig,
 )
 from vocode.streaming.transcriber.base_transcriber import (
     BaseAsyncTranscriber,
     Transcription,
 )
 from vocode.streaming.utils.worker import AsyncWorker
 from websockets.client import WebSocketClientProtocol
+from numpy.typing import DTypeLike
+from pydantic import BaseModel
 
 PUNCTUATION_TERMINATORS = [".", "!", "?"]
 MAX_SILENCE_DURATION = 2.0
 NUM_RESTARTS = 5
 
 # Silero VAD setup
+pp = PrettyPrinter(indent=2, width=80)
 model, utils = torch.hub.load(repo_or_dir="snakers4/silero-vad", model="silero_vad")
 (get_speech_timestamps, save_audio, read_audio, VADIterator, collect_chunks) = utils
 
-# Constants for Silero VAD
-USE_INT16 = False
-SAMPLE_RATE = 16000 if USE_INT16 else 8000
-DTYPE = np.int16 if USE_INT16 else np.int8
-CHUNK = 512 if USE_INT16 else 256
-WINDOWS = 3
-WINDOW_SIZE = WINDOWS * SAMPLE_RATE // CHUNK
-PREFIX_PADDING_MS = 150  # Minimum speech duration to trigger detection
-GRACE_PERIOD_MS = 300  # Grace period before resetting speech detection
 
+class AudioEncodingModel(BaseModel):
+    name: str
+    vad_chunk_size: int
+    dtype: DTypeLike
+    vad_sampling_rate: int
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+ENCODING_MAPPING = {
+    AudioEncoding.LINEAR16: AudioEncodingModel(
+        name="linear16",
+        dtype=np.int16,
+        vad_chunk_size=1024,  # increase min size because I hit he error where vad chunk size was too small
+        vad_sampling_rate=16_000,  # recomended by google speech to text
+    ),
+    AudioEncoding.MULAW: AudioEncodingModel(
+        name="mulaw",
+        dtype=np.int8,
+        vad_chunk_size=256,
+        vad_sampling_rate=8_000,
+    ),
+}
+
+
+ConfigType = TypeVar("ConfigType", contravariant=True, bound=TranscriberConfig)
```

Comment:
This is just a type anotation and it's the one which silences the editor errors, contravariant just mean that it going to be wrapped in a container class, otherwise you would have to add another type to the class generic and after trying that approach it propagates everywhere

Reply:
https://stackoverflow.com/a/61737894
![image](https://github.com/user-attachments/assets/035b6834-a43b-47d8-8ff3-bcfaecf28ef4)

Reply:
It has a fancy name because it's from math, but it just when you want to propagate inheritance to the container classes (generics).